### DISCUSSION TITLE
Can't get Mikro orm work with ES6 module

### DISCUSSION THREAD

**Discussion by ithilbor** [mikro-orm/mikro-orm#3068] [Q&A] [answered]
**Describe the bug**
I'm trying to use Mikro-ORM with the ES6 module bug in this process while i was following your documentation online it gives me an error during the init process, the error was:

Type instantiation is excessively deep and possibly infinite .ts (2589)

The error is reported by this line of code:

const orm = await MikroORM.init<PostgreSqlDriver> --> if don't specify the Driver it dosen't throw the error but i don't think it's a good way to proceed.

**Stack trace**
Using a command with the mikro-orm-cli stack trace error:
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /../server/configs/mikro-orm.config.ts
require() of ES modules is not supported.
require() of /../server/configs/mikro-orm.config.ts from /../server/node_modules/@mikro-orm/core/utils/Utils.js is an ES module file as it is a .ts file whose nearest parent package.json contains "type": "module" which defines all .ts files in that package scope as ES modules.
Instead change the requiring code to use import(), or remove "type": "module" from /../server/package.json.

Another sta trace that i get using the command yarn dev is:
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/../server/dist/models/test-entity' imported from /../server/dist/server.js

**To Reproduce**
1. Install an express + nodejs + typescript server and follow the official guide

**Expected behavior**
To work without any issue

**Additional context**
My actual package.json:
{
  "name": "test",
  "version": "1.0.0",
  "description": "test",
  "type": "module",
  "main": "server.ts",
  "repository": "test",
  "author": "test",
  "license": "GPL-3.0-or-later",
  "private": true,
  "dependencies": {
    "@mikro-orm/core": "^5.1.2",
    "@mikro-orm/mysql": "^5.1.2",
    "cors": "^2.8.5",
    "dotenv": "^16.0.0",
    "express": "^4.17.3"
  },
  "scripts": {
    "build": "yarn tsc",
    "dev": "concurrently \"yarn tsc --watch\" \"nodemon -q dist/server.js\"",
    "start": "node ./dist/server.js"
  },
  "devDependencies": {
    "@mikro-orm/cli": "^5.1.2",
    "@types/express": "^4.17.13",
    "@types/node": "^17.0.25",
    "concurrently": "^7.1.0",
    "nodemon": "^2.0.15",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.3"
  },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./configs/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
}

My tsconfig.js:
{
  "compilerOptions": {
  "target": "es6",
    "lib": [
      "esnext"
    ],
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "module": "es6",
    "moduleResolution": "node",
    "sourceMap": true, 
    "outDir": "dist",
    "removeComments": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}

**Versions**
Mac OS 11.6.5
other in package.json

**Comment by B4nan** [comment, marked as answer]
> Type instantiation is excessively deep and possibly infinite .ts (2589)

Several people reported they saw this inside vscode, not during compilation, but just inside vscode, and it was because of using dev build of TS 4.7 inside vscode. Maybe that is your case?

Note that your tsconfig seems wrong to me, you definitely don't want to target `es6` (aka ES2015) if you want to use ES modules - those are supported only in . You need esnext or something like that, this setup works for me: 

```
"compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    ...
},
```

> Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /../server/configs/mikro-orm.config.ts

You need to set `MIKRO_ORM_DYNAMIC_IMPORTS` env var to some truthy value to use dynamic imports. 

You can see working setup of express with ES modules here:

https://github.com/mikro-orm/express-js-example-app

It does not use TS, but should give you more ideas on what's needed in terms of ES modules. More about TS specific constraints are here:

https://www.typescriptlang.org/docs/handbook/esm-node.html

**Reply by ithilbor** [reply]
Hi Martin, thanks for your fast replay and yep i've checked and i've left the wrong module and target but know that i've change these two and i've also set the env var `MIKRO_ORM_DYNAMIC_IMPORTS` to 1 , the command `yarn dev` gives the same error:

[1] Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/../server/dist/models/test-entity' imported from /../server/dist/server.js
[1]     at new NodeError (node:internal/errors:371:5)
[1]     at finalizeResolution (node:internal/modules/esm/resolve:418:11)
[1]     at moduleResolve (node:internal/modules/esm/resolve:983:10)
[1]     at defaultResolve (node:internal/modules/esm/resolve:1080:11)
[1]     at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
[1]     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
[1]     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
[1]     at link (node:internal/modules/esm/module_job:78:36) {
[1]   code: 'ERR_MODULE_NOT_FOUND'
[1] }

i've also checked that these two paths are correct and they are correct, what can i miss?
Thanks in advance.

**Reply by ithilbor** [reply]
And yes i'm using JavaScript and TypeScript Nightly v4.7.20220425, now i'm going to uninstall it and reload vscode to se if the error disappear

**Reply by B4nan** [reply]
You need to use extensions in your imports.

**Reply by ithilbor** [reply]
Confirming that it was the TS extension that was throwing that error, thanks, i will keep that in mind.
Now i'm checking for understand better why it continues to give this error.
Can you also suggest me if it is good to use the ES module for an application that needs to be taken into production or is too early to use the ES module?
Thanks.

**Reply by ithilbor** [reply]
Yep, it was the extension missing, noobie errors, sorry.

**Reply by B4nan** [reply]
I am staying on commonjs for libraries but trying out es modules in end projects. Things are a bit different (like different ways to get current directory from the file, as things like `__dirname` are not available in es modules), but in general its quite usable nowadays. There might be dependencies that require es modules in latest versions, as well as others that won't work with them (or will require additional setup, like we do here), so it mainly depends on what dependencies you plan on using.

**Reply by ithilbor** [reply]
Clear, i will check for the compatibility for each library that i think i'm going to use in my project and think whats the best thing to use.
Another answer that i have is:
I decided to use .ts entities and .ts confg file as i saw on the guide i set up my .env file which is:
MIKRO_ORM_DYNAMIC_IMPORTS=1
MIKRO_ORM_DEBUG=true
MIKRO_ORM_CLI=./configs/mikro-orm.config.ts
MIKRO_ORM_CLI_USE_TS_NODE=1
MIKRO_ORM_CLI_TS_CONFIG_PATH=./tsconfig.json
and my tsconfig.json now is:
{
  "compilerOptions": {
    "target": "esnext", 
    "lib": [
      "esnext"
    ],
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "module": "esnext",
    "moduleResolution": "node",
    "declaration": true,
    "outDir": "dist",
    "removeComments": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
And my package.json:
{
  "name": "test",
  "version": "1.0.0",
  "description": "test",
  "type": "module",
  "main": "server.ts",
  "repository": "test",
  "author": "test",
  "license": "GPL-3.0-or-later",
  "private": true,
  "dependencies": {
    "@mikro-orm/core": "^5.1.2",
    "@mikro-orm/mysql": "^5.1.2",
    "@mikro-orm/reflection": "^5.1.2",
    "cors": "^2.8.5",
    "dotenv": "^16.0.0",
    "express": "^4.17.3"
  },
  "scripts": {
    "build": "yarn tsc",
    "dev": "concurrently \"yarn tsc --watch\" \"nodemon -q dist/server.js\"",
    "start": "node ./dist/server.js"
  },
  "devDependencies": {
    "@mikro-orm/cli": "^5.1.2",
    "@types/express": "^4.17.13",
    "@types/node": "^17.0.25",
    "concurrently": "^7.1.0",
    "nodemon": "^2.0.15",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.3"
  },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./configs/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
}
But when i'm using the import module like: import { Test } from "../models/test-entity";
It continue to give me the [1] Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/../server/dist/models/test-entity' imported from /../server/dist/server.js
I understood that he needs the extension in fact if a use the .js extension it works but there isn't a smarter way to let the build command put automatically that .js extension for me, for example by using a setting in tsconfig.json?
Thanks again!

**Reply by B4nan** [reply]
> but there isn't a smarter way to let the build command put automatically that .js extension for me, for example by using a setting in tsconfig.json?

I already told you, you _need_ to use extensions with ESM. And it needs to be `.js`, not `.ts`, even in the TS files. This is even noted in the guide I gave you link to. ESM modules are full of surprises like this, please read first about it (I already gave you a link to the TS docs, this is all documented there). 

This is not the right place to discuss such things, as it has nothing to do with MikroORM - I can't give you personal lectures on ES modules, not for free... :]

**Reply by ithilbor** [reply]
Yep i saw that, i was wondering if there was another way to avoid the .js specification but i think now i need to import them this way, maybe in the future nodeJS version it will not be necessary, thanks a lot for everything!

**Reply by micchickenburger** [reply]
You can run `NODE_OPTIONS=--es-module-specifier-resolution=node npx mikro-orm migration:create` to avoid needing to specify extensions in imports.
